[BUG] compiler crash using go 1.6.x

hi, it's me again and again....

## how to reproduce
I tried to compile jsonnet on a brand new vm on gce, using the ubuntu 16.04 and n1-highcpu-2（2  vCPU,1.8 GB）
I downloaded the go binary from https://dl.google.com/go/go1.6.4.linux-amd64.tar.gz and install following the instruction at https://golang.org/doc/install.
And i executed `go get -u github.com/fatih/color`, then `go build`.

![image](https://user-images.githubusercontent.com/3314335/34975848-bdaee99c-face-11e7-8468-99b1e40ab3a7.png)

![image](https://user-images.githubusercontent.com/3314335/34975855-c3ae8654-face-11e7-9655-5d9275b585a2.png)

![image](https://user-images.githubusercontent.com/3314335/34975862-cae5106e-face-11e7-84b1-2790d6ccdecf.png)

I tested on a larger vm, and it turns out that there may be a bug in go 1.6 compiler so it's not able to compile using 1.6, but 1.7.0 and 1.7.6 works fine.
```
daniel@instance-2:~$ sudo tar -C /usr/local -xzf go1.6.4.linux-amd64.tar.gz
daniel@instance-2:~$ export PATH=$PATH:/usr/local/go/bin
daniel@instance-2:~$ mkdir -p go/src/github.com/google
daniel@instance-2:~$ cd go/src/github.com/google/
daniel@instance-2:~/go/src/github.com/google$ git clone https://github.com/google/go-jsonnet.git
Cloning into 'go-jsonnet'...
remote: Counting objects: 1839, done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 1839 (delta 6), reused 19 (delta 3), pack-reused 1802
Receiving objects: 100% (1839/1839), 5.28 MiB | 1.51 MiB/s, done.
Resolving deltas: 100% (698/698), done.
Checking connectivity... done.
daniel@instance-2:~/go/src/github.com/google$ cd go-jsonnet/
daniel@instance-2:~/go/src/github.com/google/go-jsonnet$ ls
ast          compat        desugarer.go       dump                imports.go           jsonnet          linter                             parser              runtime_error.go         std                   tests.sh   vm.go
builtins.go  compat_test   desugarer_test.go  error_formatter.go  interpreter.go       jsonnet_test.go  main_test.go                       README.md           static_analyzer.go       std.thisFile.jsonnet  thunks.go
cmd          CONTRIBUTING  doc.go             evaluator.go        interpreter_test.go  LICENSE          mutually_recursive_defaults.input  reset_stdast_go.sh  static_analyzer_test.go  testdata              value.go
daniel@instance-2:~/go/src/github.com/google/go-jsonnet$ pwd
/home/daniel/go/src/github.com/google/go-jsonnet
daniel@instance-2:~/go/src/github.com/google/go-jsonnet$ ls
ast          compat        desugarer.go       dump                imports.go           jsonnet          linter                             parser              runtime_error.go         std                   tests.sh   vm.go
builtins.go  compat_test   desugarer_test.go  error_formatter.go  interpreter.go       jsonnet_test.go  main_test.go                       README.md           static_analyzer.go       std.thisFile.jsonnet  thunks.go
cmd          CONTRIBUTING  doc.go             evaluator.go        interpreter_test.go  LICENSE          mutually_recursive_defaults.input  reset_stdast_go.sh  static_analyzer_test.go  testdata              value.go
daniel@instance-2:~/go/src/github.com/google/go-jsonnet$ pwd
/home/daniel/go/src/github.com/google/go-jsonnet
daniel@instance-2:~/go/src/github.com/google/go-jsonnet$ cd jsonnet/
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go build
cmd.go:31:2: cannot find package "github.com/fatih/color" in any of:
	/usr/local/go/src/github.com/fatih/color (from $GOROOT)
	($GOPATH not set)
cmd.go:32:2: cannot find package "github.com/google/go-jsonnet" in any of:
	/usr/local/go/src/github.com/google/go-jsonnet (from $GOROOT)
	($GOPATH not set)
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go get -u github.com/fatih/color
package github.com/fatih/color: cannot download, $GOPATH not set. For more details see: go help gopath
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ export GOPATH=$HOME/go
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go get -u github.com/fatih/color
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go build
# github.com/google/go-jsonnet/ast
panic: runtime error: makeslice: len out of range

goroutine 1 [running]:
panic(0x803ba0, 0xc84dd40b70)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
cmd/compile/internal/gc.newliveness(0xc82a2626c0, 0xc82af6cfc0, 0xc8518a6000, 0x3cd44, 0x48800, 0xc851ee4000, 0xd09c, 0xf000, 0x9869ac5bda4f6ea8)
	/usr/local/go/src/cmd/compile/internal/gc/plive.go:687 +0x161
cmd/compile/internal/gc.liveness(0xc82a2626c0, 0xc82af6cfc0, 0xc833fc9380, 0xc833fc9400)
	/usr/local/go/src/cmd/compile/internal/gc/plive.go:1782 +0x2cf
cmd/compile/internal/gc.compile(0xc82a2626c0)
	/usr/local/go/src/cmd/compile/internal/gc/pgen.go:541 +0xdf2
cmd/compile/internal/gc.funccompile(0xc82a2626c0)
	/usr/local/go/src/cmd/compile/internal/gc/dcl.go:1450 +0x1c0
cmd/compile/internal/gc.fninit(0xc82293c4a0)
	/usr/local/go/src/cmd/compile/internal/gc/init.go:184 +0x8a4
cmd/compile/internal/gc.Main()
	/usr/local/go/src/cmd/compile/internal/gc/lex.go:477 +0x2150
cmd/compile/internal/amd64.Main()
	/usr/local/go/src/cmd/compile/internal/amd64/galign.go:127 +0x58d
main.main()
	/usr/local/go/src/cmd/compile/main.go:32 +0x395
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ cd
daniel@instance-2:~$ https://dl.google.com/go/go1.7.6.linux-amd64.tar.gz
-bash: https://dl.google.com/go/go1.7.6.linux-amd64.tar.gz: No such file or directory
daniel@instance-2:~$ wget https://dl.google.com/go/go1.7.6.linux-amd64.tar.gz
--2018-01-16 07:35:25--  https://dl.google.com/go/go1.7.6.linux-amd64.tar.gz
Resolving dl.google.com (dl.google.com)... 74.125.68.91, 74.125.68.93, 74.125.68.190, ...
Connecting to dl.google.com (dl.google.com)|74.125.68.91|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 84237339 (80M) [application/x-gzip]
Saving to: ‘go1.7.6.linux-amd64.tar.gz’

go1.7.6.linux-amd64.tar.gz                                 100%[========================================================================================================================================>]  80.33M   221MB/s    in 0.4s

2018-01-16 07:35:25 (221 MB/s) - ‘go1.7.6.linux-amd64.tar.gz’ saved [84237339/84237339]

daniel@instance-2:~$ sudo rm -rf /usr/local/go/
daniel@instance-2:~$ sudo tar -C /usr/local -xzf go1.7.6.linux-amd64.tar.gz
daniel@instance-2:~$ go version
go version go1.7.6 linux/amd64
daniel@instance-2:~$ cd go/src/github.com/google/go-jsonnet/jsonnet/
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go build
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ free -m
              total        used        free      shared  buff/cache   available
Mem:          15032         155       13973           8         903       14556
Swap:             0           0           0
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ cd
daniel@instance-2:~$ sudo rm -rf /usr/local/go/
daniel@instance-2:~$ wget https://dl.google.com/go/go1.7.linux-amd64.tar.gz
--2018-01-16 07:37:09--  https://dl.google.com/go/go1.7.linux-amd64.tar.gz
Resolving dl.google.com (dl.google.com)... 74.125.200.93, 74.125.200.190, 74.125.200.136, ...
Connecting to dl.google.com (dl.google.com)|74.125.200.93|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 81573766 (78M) [application/x-gzip]
Saving to: ‘go1.7.linux-amd64.tar.gz’

go1.7.linux-amd64.tar.gz                                   100%[========================================================================================================================================>]  77.79M  37.3MB/s    in 2.1s

2018-01-16 07:37:12 (37.3 MB/s) - ‘go1.7.linux-amd64.tar.gz’ saved [81573766/81573766]

daniel@instance-2:~$ sudo tar -C /usr/local -xzf go1.7.linux-amd64.tar.gz
daniel@instance-2:~$ cd go/src/github.com/google/go-jsonnet/jsonnet/
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ go build
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$ free -m
              total        used        free      shared  buff/cache   available
Mem:          15032         157       13893           8         981       14552
Swap:             0           0           0
daniel@instance-2:~/go/src/github.com/google/go-jsonnet/jsonnet$
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,6 +2,7 @@ language: go
 sudo: false
 go:
   - 1.x
+  - 1.8.x
   - tip
 before_install:
   - go get github.com/axw/gocov/gocov

--- a/README.md
+++ b/README.md
@@ -14,7 +14,7 @@ feature complete but is not as heavily exercised as the [Jsonnet C++
 implementation](https://github.com/google/jsonnet).  Please try it out and give
 feedback.
 
-This code is known to work on Go 1.6 and above.
+This code is known to work on Go 1.8 and above. We recommend always using the newest stable release of Go.
 
 ## Build instructions
 

--- a/builtins.go
+++ b/builtins.go
@@ -549,7 +549,14 @@ var builtinAsin = liftNumeric(math.Asin)
 var builtinAcos = liftNumeric(math.Acos)
 var builtinAtan = liftNumeric(math.Atan)
 var builtinLog = liftNumeric(math.Log)
-var builtinExp = liftNumeric(math.Exp)
+var builtinExp = liftNumeric(func(f float64) float64 {
+	res := math.Exp(f)
+	if res == 0 && f > 0 {
+		return math.Inf(1)
+	} else {
+		return res
+	}
+})
 var builtinMantissa = liftNumeric(func(f float64) float64 {
 	mantissa, _ := math.Frexp(f)
 	return mantissa

